lib/std/zig/system/NativeTargetInfo.zig

```diff
@@ -355,37 +355,63 @@ fn detectAbiAndDynamicLinker(
-        error.NameTooLong => unreachable,
-        error.PathAlreadyExists => unreachable,
-        error.SharingViolation => unreachable,
-        error.InvalidUtf8 => unreachable,
-        error.BadPathName => unreachable,
-        error.PipeBusy => unreachable,
-        error.FileLocksNotSupported => unreachable,
-        error.WouldBlock => unreachable,
-        error.FileBusy => unreachable, // opened without write permissions
-
-        error.IsDir,
-        error.NotDir,
-        error.InvalidHandle,
-        error.AccessDenied,
-        error.NoDevice,
-        error.FileNotFound,
-        error.FileTooBig,
-        error.Unexpected,
-        => return defaultAbiAndDynamicLinker(cpu, os, cross_target),
+    const elf_file = blk: {
+        // This block looks for a shebang line in /usr/bin/env,
+        // if it finds one, then instead of using /usr/bin/env as the ELF file to examine, it uses the file it references instead,
+        // doing the same logic recursively in case it finds another shebang line.
+
+        // Since /usr/bin/env is hard-coded into the shebang line of many portable scripts, it's a
+        // reasonably reliable path to start with.
+        var file_name: []const u8 = "/usr/bin/env";
+        while (true) {
+            const file = std.fs.openFileAbsolute(file_name, .{}) catch |err| switch (err) {
+                error.NoSpaceLeft => unreachable,
+                error.NameTooLong => unreachable,
+                error.PathAlreadyExists => unreachable,
+                error.SharingViolation => unreachable,
+                error.InvalidUtf8 => unreachable,
+                error.BadPathName => unreachable,
+                error.PipeBusy => unreachable,
+                error.FileLocksNotSupported => unreachable,
+                error.WouldBlock => unreachable,
+                error.FileBusy => unreachable, // opened without write permissions
+
+                error.IsDir,
+                error.NotDir,
+                error.InvalidHandle,
+                error.AccessDenied,
+                error.NoDevice,
+                error.FileNotFound,
+                error.FileTooBig,
+                error.Unexpected,
+                => return defaultAbiAndDynamicLinker(cpu, os, cross_target),
 
-        else => |e| return e,
+                else => |e| return e,
+            };
+            var should_close = true;
+            defer if (should_close) file.close();
+
+            // #! (2) + 255 (max length of shebang line since Linux 5.1) + \n (1)
+            var buffer: [258]u8 = undefined;
+            const line = file.reader().readUntilDelimiter(&buffer, '\n') catch {
+                should_close = false;
+                break :blk file;
+            };
+            if (mem.startsWith(u8, line, "#!")) {
+                var it = std.mem.tokenize(u8, line[2..], " ");
+                file_name = it.next() orelse return defaultAbiAndDynamicLinker(cpu, os, cross_target);
+                continue;
```

Comment:
`file_name` is a pointer into `buffer`, which is invalidated at this point

Reply:
> I think there's a memory-validity problem here: `file_name` is a pointer into `buffer`, which is invalidated at this point
> 
> You should be able to fix it by changing where `buffer` is declared

This may sounds strange, but... this is the reason why I changed where `buffer` is declared...

Reply:
I think it should be placed outside of the while loop - Do you agree?

Reply:
> I think it should be placed outside of the while loop - Do you agree?

I'll check this in the morning :)

Reply:
> I think it should be placed outside of the while loop - Do you agree?

Done!